### ISSUE TITLE
Atom reordering does not correctly occur on CUDA/OpenCL platforms after setPositions() is called

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#3705]
We have discovered that calling `setPositions()` calls [`reorderAtoms()`](https://github.com/openmm/openmm/blob/109f6b2535da4e0c0dd88007d6ca06b4add2ce81/platforms/cuda/src/CudaKernels.cpp#L221) to optimally reorder atoms for efficiency, but this does not actually cause atoms to be reordered because of a check for [`stepsSinceReorder < 250`](https://github.com/openmm/openmm/blob/3d62421b62494d3d3143feedf0d580c74501f92f/platforms/common/src/ComputeContext.cpp#L435). 

This seems like unintended behavior, and that adding `cu.setStepsSinceReorder(99999)` right before [cu.reorderAtoms()](https://github.com/openmm/openmm/blob/109f6b2535da4e0c0dd88007d6ca06b4add2ce81/platforms/cuda/src/CudaKernels.cpp#L221) would ensure the atoms are correctly reordered after `setPositions()` is called.

cc https://github.com/choderalab/perses/issues/613#issuecomment-1189514739

I can open a PR to fix this.

**Comment by peastman** [comment]
This behavior is intended.  Actually reordering atoms is very expensive.  We treat a call to `setPositions()` as equivalent to taking a step of dynamics: we have to assume successive calls will set generally similar positions, but if you're doing it a lot we should occasionally reorder.  Actually reordering on every call would be ruinously slow in applications where you repeatedly call `setPositions()` to evaluate a series of conformations.

**Comment by jchodera** [comment]
What do you propose as an alternative? Could we add some kind of API call that would enable us to force atom reordering?

**Comment by jchodera** [comment]
As is seen here, we have quantitative evidence that _not_ reordering is ruinously slow for replica exchange simulations where multiple trajectories are sequenced serially:
![image](https://user-images.githubusercontent.com/3656088/179883702-a44e5a00-b1fc-467b-8997-9b99f9cef41f.png)
https://github.com/choderalab/perses/issues/613#issuecomment-1189514739

**Comment by jchodera** [comment]
> Actually reordering on every call would be ruinously slow in applications where you repeatedly call setPositions() to evaluate a series of conformations.

We could add an optional argument to `setPositions()` like `reorderAtoms=False` that can be set to `True` to preserve existing behavior but allow for applications to get better performance if reordering is needed.

**Comment by jchodera** [comment]
Alternatively, we could reorder atoms in the call to `Integrator.step()` if we have set a flag that we just called `setPositions()`. That would automatically do the right thing.

**Comment by jchodera** [comment]
We tested the impact on speed if we're just calling `setPositions()` followed by `context.getState(getEnergy=True).getPotentialEnergy()`. There is indeed an impact on performance.

On an integrated GPU:
* With fix: 500 cycles of of setPositions + energy evaluations took 0.493 seconds
* Without it: 500 cycles of of setPositions + energy evaluations took 0.369 seconds
On an eGPU A4000:
* with fix: 500 cycles of of setPositions + energy evaluations took 0.445 seconds
* Without fix: 500 cycles of of setPositions + energy evaluations took 0.257 seconds

<details>
<summary>Here's the script we used:</summary>

```python
from simtk import unit, openmm
from openmmtools import testsystems

t = testsystems.AlchemicalWaterBox()
system, positions = t.system, t.positions

from openmmtools.utils import get_fastest_platform
platform = get_fastest_platform(minimum_precision='mixed')
integrator = openmm.VerletIntegrator(0.001)
context = openmm.Context(system, integrator, platform)
context.setPositions(positions)

# Take a step to compile kernels                                                                                                                                                                                                                                                 
integrator.step(1)
context.getState(getEnergy=True).getPotentialEnergy()

# Timing cycles
unitless_positions = positions.value_in_unit_system(unit.md_unit_system)
import time
ncycles = 500
initial_time = time.time()
for cycle in range(ncycles):
    context.setPositions(unitless_positions)
    context.getState(getEnergy=True).getPotentialEnergy()
elapsed_time = time.time() - initial_time
print(f'{ncycles} cycles of of setPositions + energy evaluations took {elapsed_time:.3f} seconds')
```
</details>
This suggests it would indeed be better to trigger atom reordering before the first step of dynamics, and thereafter count number of dynamics steps since last atom reordering.

**Comment by jchodera** [comment]
@peastman : Do you have an alternative proposal I could implement?

**Comment by peastman** [comment]
An alternative method is implemented in #3721.
